app/models/book.rb

```diff
@@ -1,3 +1,60 @@
 class Book < ApplicationRecord
   has_and_belongs_to_many :tags
+  has_one :reservation, dependent: :destroy
+  belongs_to :user, foreign_key: "user_id"
+
+  validate :cannot_edit_reserved_or_booked
+  before_destroy :cannot_destroy_reserved_or_booked
+
+  STATUSES = [
+    AVAILABLE = 'available',
+    RESERVED = 'reserved',
+    BORROWED = 'borrowed'
+  ]
+
+  validates :status, :inclusion => { :in => STATUSES }, :allow_blank => false
+
+  def belongs_to?(current_user)
+  	user == current_user
+  end
+
+  def is_available?
+    status == AVAILABLE
+  end
+
+  def available_to_borrow?(current_user)
+  	status == AVAILABLE && !belongs_to?(current_user)
+  end
+
+  def reserved_by?(current_user)
+    status == RESERVED && reservation.user == current_user
+  end
+
+  def borrower
+    reservation.user
+  end
+
+  def has_pending_reservation?
```

Comment:
there's no concept of `pending` in our states, I feel like this naming is a bit overcomplicating things, can we just call it `has_reservation` or `is_reserved`?

Reply:
Makes sense, at that stage I was clearly still figuring out the best names for things -- I've now changed this to `is_reserved?`